Feat : Highlight triggered and errored blocks

abdou6666 opened the pull request:
# Motivation
[Screencast from 07-04-25 13:03:35.webm](https://github.com/user-attachments/assets/319a82d5-3696-4c16-ace4-09d47c50ca55)

**Before testing you might need to do `hexabot stop hexabot dev`**
This PR introduces the feature of highlighting trigger blocks in the visual editor. Additionally, it adds error highlighting for plugins when they encounter issues, improving the visibility and debugging experience for developers.

Fixes #884

# Type of change:

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have performed a self-review of my own code

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Enhanced error tracking and live notifications during message processing provide clearer user feedback.
  - Upgraded real-time interactions through WebSocket, enabling dynamic highlights and error notifications in the visual editor.
  - Improved visual editor navigation with functions that auto-center blocks and enable smooth transitions between flows.
  - Added new event handling for block highlighting and error scenarios, improving user interaction.
  - Introduced new state management for node highlighting, error, and fallback statuses in the visual editor.
  - Integrated WebSocket query mechanism in the Diagrams component for real-time updates.
  - New methods for managing highlighted, errored, and fallback states in the visual editor's node model.
  - Added WebSocket subscription capabilities for user-specific interactions.
  - Added a new chatbot setting to enable flow debugging, which highlights executed, failed, or fallback blocks during flow runs for real-time visual debugging.

- **Style**
  - Refined visual cues and animations for highlighted, error, and fallback states ensure a consistent and engaging user experience.

- **Chores**
  - Various internal enhancements and cleanup have been made to boost overall stability and performance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

medchedli commented:
I'm noticed this error shows up in the API logs: `TypeError: Cannot read properties of null (reading 'id')` originating from triggerBlock

This error occurs when handling a local fallback (user provides invalid input). It seems the block object passed to triggerBlock in this fallback scenario has block.category set to null. The code then tries to access block.category.id, leading to the error.

```
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Conversation has been captured! Responding ...
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BlockService] Matched patterns: []
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Responding ...
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] 680a479a02a1c4b93d21bd3f
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Respond to nested conversion! Go next 
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] 5efd9c51a518c9444f737177
[Nest] 2065  - 04/24/2025, 2:16:27 PM   ERROR [BotService] Unable to process/send message.
[Nest] 2065  - 04/24/2025, 2:16:27 PM   ERROR [BotService] TypeError: Cannot read properties of null (reading 'id')
    at BotService.triggerBlock (/app/src/chat/services/bot.service.ts:155:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at BotService.handleIncomingMessage (/app/src/chat/services/bot.service.ts:328:11)
    at BotService.processConversationMessage (/app/src/chat/services/bot.service.ts:386:14)
    at BotService.handleMessageEvent (/app/src/chat/services/bot.service.ts:484:24)
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [ChatService] Conversation has ended successfully.
[Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [ChatService] 680a479a02a1c4b93d21bd3f
```

abdou6666 commented:
> I'm noticed this error shows up in the API logs: `TypeError: Cannot read properties of null (reading 'id')` originating from triggerBlock
> 
> This error occurs when handling a local fallback (user provides invalid input). It seems the block object passed to triggerBlock in this fallback scenario has block.category set to null. The code then tries to access block.category.id, leading to the error.
> 
> ```
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Conversation has been captured! Responding ...
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BlockService] Matched patterns: []
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Responding ...
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] 680a479a02a1c4b93d21bd3f
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] Respond to nested conversion! Go next 
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [BotService] 5efd9c51a518c9444f737177
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   ERROR [BotService] Unable to process/send message.
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   ERROR [BotService] TypeError: Cannot read properties of null (reading 'id')
>     at BotService.triggerBlock (/app/src/chat/services/bot.service.ts:155:33)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at BotService.handleIncomingMessage (/app/src/chat/services/bot.service.ts:328:11)
>     at BotService.processConversationMessage (/app/src/chat/services/bot.service.ts:386:14)
>     at BotService.handleMessageEvent (/app/src/chat/services/bot.service.ts:484:24)
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [ChatService] Conversation has ended successfully.
> [Nest] 2065  - 04/24/2025, 2:16:27 PM   DEBUG [ChatService] 680a479a02a1c4b93d21bd3f
> ```

Hi @medchedli , I fixed this issue and i added a setting that enables/disables this feature as you recommended !
